### ISSUE TITLE
Stop setting $PAGER by default

### ISSUE THREAD

**Issue by leroycep** [NixOS/nixpkgs#361613]
## Describe the bug

NixOS is setting `PAGER` to `less` by default. I would prefer it not be set unless I explicitly ask for it.

## Steps To Reproduce

```source
$ env -i bash
bash-5.2$ echo $PAGER
less 
bash-5.2$ 
```

## Expected behavior

```source
$ env -i bash
bash-5.2$ echo $PAGER
bash-5.2$ 
```

## Additional context

I've been annoyed with [`jujutsu`](https://martinvonz.github.io/jj/latest/) for a bit because it would always open up the log in a temporary buffer. When I looked at the settings, I discovered that it was defaulted to passing some flags to `less` that make it way better to use. Neat! Now why wasn't I getting those flags?

The end of rabbit hole was these lines here:

https://github.com/NixOS/nixpkgs/blob/a2283c7a2e280fbaae9f9fec8cfa0e6259e0e056/nixos/modules/programs/environment.nix#L17-L23

## Metadata

```console
nu ❯ nix-shell -p nix-info --run "nix-info -m"
this path will be fetched (0.02 MiB download, 0.06 MiB unpacked):
  /nix/store/m1p78gqlc0pw3sdbz3rdhklzm0g26g96-stdenv-linux
copying path '/nix/store/m1p78gqlc0pw3sdbz3rdhklzm0g26g96-stdenv-linux' from 'https://cache.nixos.org'...
 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.63, NixOS, 24.11 (Vicuna), 24.11.20241130.62c435d`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.24.10`
 - channels(root): `"nixos-22.05"`
 - channels(geemili): `""`
 - nixpkgs: `/nix/store/ljk3wvwqqzbmz0pys49ii18qmsi5aibc-source`
```

## Notify maintainers

Not really sure who to @

- The person who originally made PAGER have a default value, as determined by `git log -S PAGER --oneline --reverse`: @edolstra 
- last person to touch the specific line in `git blame`, though the change was unrelated: @acid-bong

---

Note for maintainers: Please tag this issue in your PR.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by thefossguy** [comment]
These are sane defaults. If you want something else, you should put it in your shell's configuration file.

**Comment by leroycep** [comment]
Yeah, I don't necessarily disagree. I was a bit frustrated when I made this issue, maybe not in the best state of mind.

That said, I still don't understand why `$PAGER` is being set. Isn't `$PAGER` being set meant to indicate that the user has made a choice? Clearly at least `jujutsu` thinks it has been set by a user.

Basically, I'm confused and trying to figure out the following:

1. Does `$PAGER` need to be set? Does something break without it set?
2. Is `$PAGER` supposed to be a user configuration setting? (That is to say, is this a bug on `jujutsu`'s side for assuming `$PAGER` represents the user declared preference?)
3. Is there a way to unset the `$PAGER` from the NixOS configuration? I haven't actually done a deep dive on terminal pagers, and I would prefer it stay that way to be honest. But it's more annoying when `$PAGER` is saying that I specifically want paging done with `less`.

As some additional context, I discovered this when I recently installed Arch Linux and realized that `jujutsu`'s paging was much improved.

**Comment by acid-bong** [comment]
You can always make smth like `alias jj="PAGER= jj"`

**Comment by leroycep** [comment]
That is not the problem I am trying to report here. I have already added `hide-env PAGER` to my nushell config. The problem is that `jj` was assuming `PAGER` to be user configuration, and NixOS is treating `PAGER` as a default, fallback option. NixOS and `jj` are using the `PAGER` variable for two conflicting things.

What I'm trying to figure out here is how to resolve these conflicting definitions. In my (relatively uninformed) opinion, `jj`'s usage of `PAGER` makes more sense, and I don't understand why NixOS is using it the way it is.

**Comment by pluiedev** [comment]
My 2¢ would be that PAGER is akin to EDITOR: it is an agreed-upon (not necessarily standardized or codified) environment variables that users may set to indicate what pager they'd like to use, but it should default to *something* — there might be software that don't fail gracefully without a pager, and I can't name a single example aside from Jujutsu that has its own builtin pager.

Although, unlike EDITOR, there's no module options AFAICT that overrides the default PAGER setting, which IMO is surprising behavior (and our usage of less is not without bugs at times (see #354377), so I think that can be one area of improvement

**Comment by leroycep** [comment]
> My 2¢ would be that PAGER is akin to EDITOR: it is an agreed-upon (not necessarily standardized or codified) environment variable that users may set to indicate what pager they'd like to use, but it should default to something — there might be software that fail less-than-gracefully without a pager, 

It would be nice to get some concrete examples of software that fails less-than-gracefully without `$PAGER` being set. That's partially why I opened this issue instead of a pull request.

On the other hand, `jujutsu` provides a clear case of the opposite; software that fails _because_ `$PAGER` was _unintentionally_ set.

> and I can't name a single example aside from Jujutsu that has its own builtin pager.

For what it's worth, the improvement I saw wasn't from using jujutsu's builtin pager, it was from jujutsu using `less -FRX` instead of just `less`.

> Although, unlike EDITOR, there's no module options AFAICT that overrides the default PAGER setting, which IMO is surprising behavior (and our usage of less is not without bugs at times (see https://github.com/NixOS/nixpkgs/issues/354377), so I think that could be one area of improvement

Maybe something like `disable-default-pager-environment-variable`? Though setting and unsetting environment variables isn't particularly difficult usually, and my contention here is more about the unexpected default that led to a degraded user experience.

> It doesn't need to be set but it is helpful when the output is longer than the height of console.

Yes, that's why terminal pagers exist. I don't doubt that.

> As for why it doesn't need to be set and is still set regardless is because that's the standard practice, even on Arch (last I used it).

`$PAGER` isn't set for me on my Arch install, but I'll admit that my Arch install may atypical, and I have no idea what is standard. I wonder if a change was made? Perhaps there is some discussion.

> You can learn about how jujutsu interprets it here: https://martinvonz.github.io/jj/latest/config/#pager

I link that in my very first post, and that brings me back to my point: Jujutsu and NixOS are using the $PAGER environment in two different ways that are incompatible with each other. Either:

1. `jujutsu`'s interpretation that `$PAGER` is a user configuration variable is wrong, and it should stop relying on `$PAGER` to implement things like `--quit-if-one-screen`, or
2. NixOS's interpretation of `$PAGER` as some kind of fallback is wrong, and it should stop setting `$PAGER` by default.

>> Is there a way to unset the `$PAGER` from the NixOS configuration?
>
> `export PAGER=` (yes, nothing on the right hand side)

That isn't unsetting it from the NixOS configuration. For comparison, setting it via `export PAGER=`:

```console
[geemili@aitxero seizer]$ export PAGER=
[geemili@aitxero seizer]$ jj
Warning: Failed to spawn pager '': No such file or directory (os error 2)
Hint: Consider using the `:builtin` pager.
@  yszvwyrk opensource@geemili.xyz 2024-11-29 14:47:03 69afd953
│  (empty) (no description set)
◆  vyykmmnn opensource@geemili.xyz 2024-11-29 14:46:31 dev git_head() 1bdaaed6
│  feat(seizer.Image): add function that detects image format at comptime
~
```

Using `unset PAGER` (equivalent to not setting it in the first place):

```console
[geemili@aitxero seizer]$ unset PAGER
[geemili@aitxero seizer]$ jj
@  yszvwyrk opensource@geemili.xyz 2024-11-29 14:47:03 69afd953
│  (empty) (no description set)
◆  vyykmmnn opensource@geemili.xyz 2024-11-29 14:46:31 dev git_head() 1bdaaed6
│  feat(seizer.Image): add function that detects image format at comptime
~
```

`man` will also fail if `$PAGER` is set to an empty string instead of being unset. It will simply dump the entirety of the man page to the terminal instead of defaulting to `less`.

**Comment by pluiedev** [comment]
Again, my issue with the entire premise of this issue is that all the examples that demonstrate setting `PAGER` would be a bad idea, comes from Jujutsu. I'm not discrediting the concerns entirely (as a Jujutsu user myself), but if this is something worth fixing *globally* then we oughta have more examples of programs that don't play nice when `PAGER` is set to a default value.

**Comment by toonn** [comment]
Git has very similar behavior, passing `FRX` unless `LESS` is set, cf. `core.pager` entry in `man git-config(1)`.

The NixOS module went from setting `PAGER='less -R'`, which I complained about some 4 years ago, to setting `PAGER=less` and `LESS=R`, which clobbered Git's default behavior, to using Less' `LESSKEYIN_SYSTEM` functionality to provide a default that will not prevent programs like Git from properly passing what they intend to be the default in the absence of affirmative action by the user to set `LESS` or `PAGER`.

I do think the problem in this case is with Jujutsu, it basically uses (in Bash terminology) `${PAGER:-less -FRX}` when it could instead use `LESS=${LESS:-FRX} ${PAGER:-less}`.
